utils/gatewayEvm.ts

```diff
@@ -0,0 +1,216 @@
+import { AbiCoder, ethers } from "ethers";
+
+import { RevertOptions, TxOptions } from "../types/contracts.types";
+import { ParseAbiValuesReturnType } from "../types/parseAbiValues.types";
+import { toHexString } from "./toHexString";
+
+/**
+ * Creates a revert data object from revert options
+ */
+export const createRevertData = (
+  revertOptions: RevertOptions
+): RevertOptions => {
+  return {
+    abortAddress: "0x0000000000000000000000000000000000000000", // not used
+    callOnRevert: revertOptions.callOnRevert,
+    onRevertGasLimit: revertOptions.onRevertGasLimit,
+    revertAddress: revertOptions.revertAddress,
+    revertMessage: toHexString(revertOptions.revertMessage),
+  };
+};
+
+/**
+ * Gateway ABI fragments for common functions
+ */
+const GATEWAY_FRAGMENTS = {
+  call: "function call(address receiver, bytes data, tuple(address revertAddress, bool callOnRevert, address abortAddress, bytes revertMessage, uint256 onRevertGasLimit) revertOptions)",
+  deposit:
+    "function deposit(address receiver, tuple(address revertAddress, bool callOnRevert, address abortAddress, bytes revertMessage, uint256 onRevertGasLimit) revertOptions)",
+  depositAndCall:
+    "function depositAndCall(address receiver, bytes data, tuple(address revertAddress, bool callOnRevert, address abortAddress, bytes revertMessage, uint256 onRevertGasLimit) revertOptions)",
+  depositAndCallERC20:
+    "function depositAndCall(address receiver, uint256 amount, address asset, bytes data, tuple(address revertAddress, bool callOnRevert, address abortAddress, bytes revertMessage, uint256 onRevertGasLimit) revertOptions)",
+  depositERC20:
+    "function deposit(address receiver, uint256 amount, address asset, tuple(address revertAddress, bool callOnRevert, address abortAddress, bytes revertMessage, uint256 onRevertGasLimit) revertOptions)",
+};
+
+/**
+ * Generates calldata for a specific gateway method
+ */
+export const generateGatewayCallData = (
+  methodName: keyof typeof GATEWAY_FRAGMENTS,
+  args: unknown[]
+): string => {
+  const fragment = GATEWAY_FRAGMENTS[methodName];
+  const iface = new ethers.Interface([fragment]);
+
+  // The method name from the fragment (before the opening parenthesis)
+  // Extract the function name from the ABI fragment by parsing the string
+  // Example: from "function deposit(address receiver, ...)" we extract "deposit"
+  const [, functionSignature] = fragment.split(" ");
+  const [functionName] = functionSignature.split("(");
+
+  try {
+    return iface.encodeFunctionData(functionName, args);
+  } catch (error) {
+    console.error(`Error encoding calldata for ${methodName}:`, error);
+    throw error;
+  }
+};
+
+/**
+ * Prepares a gateway transaction for broadcasting
+ */
+export const prepareGatewayTx = (
+  gatewayAddress: string,
```

Comment:
Right now `prepareGatewayTx` takes gateway address and value and just passes it through returning it inside the `txData` object. Since it's not being used maybe we should just pass gateway address and value directly to `broadcastGatewayTx`.

Reply:
746c1f

Reply:
f4d70316fc149c3abdba34c244998a20968fde2b